Add scanblocks RPC call

jonasschnelli opened the pull request:
The `scanblocks` RPC call allows one to get relevant blockhashes from a set of descriptors by scanning all blockfilters in a given range.

**Example:**

`scanblocks start '["addr(<bitcoin_address>)"]' 661000`
(returns relevant blockhashes for `<bitcoin_address>` from blockrange 661000->tip)

## Why is this useful?

**Fast wallet rescans**: get the relevant blocks and only rescan those via `rescanblockchain getblockheader(<hash>)[height] getblockheader(<hash>)[height])`. A future PR may add an option to allow to provide an array of blockhashes to `rescanblockchain`.

**prune wallet rescans**: (*needs additional changes*): together with a call to fetch blocks from the p2p network if they have been pruned, it would allow to rescan wallets back to the genesis block in pruned mode (relevant #15946).

**SPV mode** (*needs additional changes*): it would be possible to build the blockfilterindex from the p2p network (rather then deriving them from the blocks) and thus allow some sort of hybrid-SPV mode with moderate bandwidth consumption (related #9483)

Sjors commented:
Concept ACK

#20295 should also help with prune wallet rescans

Sjors commented on test/functional/rpc_scanblockfilters.py:
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) 2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test the scantxoutset rpc call."""
```
`scanblockfilters`

Sjors commented on test/functional/rpc_scanblockfilters.py:
```diff
@@ -0,0 +1,29 @@
+class ScanblockfiltersTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.extra_args = [["-blockfilterindex=1"]]
```
Maybe add a node without this index and test that the RPC call fails.

jonatack replied:
and `assert_raises_rpc_error` for "Unknown filtertype", "Index is not enabled for filtertype " + filtertype_name, and "Invalid stopheight"

Sjors commented on test/functional/rpc_scanblockfilters.py:
```diff
@@ -0,0 +1,29 @@
+        assert(blockhash in self.nodes[0].scanblockfilters(["addr("+addr_1+")"]))
+        blockhash_new = self.nodes[0].generate(1)[0]
+        self.log.info("Test scanblockfilters startheight...")
+        assert(blockhash not in self.nodes[0].scanblockfilters(["addr("+addr_1+")"], self.nodes[0].getblockheader(blockhash_new)['height']))
```
This is a bit hard to read, maybe start with `start_height = self.nodes[0].getblockheader(blockhash_new)['height']`

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -2284,6 +2284,108 @@ static RPCHelpMan scantxoutset()
+                        {RPCResult::Type::STR_HEX, "", "The blockhash"},
+                    }},
+                RPCExamples{
+                    HelpExampleCli("scanblockfilters", "[\"addr(bcrt1q4u4nsgk6ug0sqz7r3rj9tykjxrsl0yy4d0wwte)\"] 300000") +
```
Would be good to add an object descriptor too.

Sjors reviewed (commented):
Concept ACK

When I call without a `height` param it doesn't actually scan (documentation suggests otherwise).

I was able to find some blocks used by addresses in my own wallet, both receiving on and sending from. If anyone wants to ACK #19136 I'll test with a ranged descriptor too.

Sadly the RPC call timed out, so it didn't return the result. You may need to add an abstraction like for `rescanwallet` where one can query rescans in progress, but also obtain the result later. That might be a lot of extra code though. 

#20295 should also help with prune wallet rescans

Don't forget to drop `Conflicts` from the second commit message.

Can you also add a test for passing a descriptor object (with range)?

Other RPC methods that involve descriptors require a checksum; we should probably do that here as well.

For a followup it would be nice if descriptors are automatically expanded, so we don't need to specify a range. This comes with some headaches, see #15845

The documentation should remind the user about false positive block matches.

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -2284,6 +2284,108 @@ static RPCHelpMan scantxoutset()
+            next = ChainActive().Next(block);
+        }
+        if (start_index->nHeight+amount_per_chunk == block->nHeight || next == nullptr) {
+            LogPrint(BCLog::RPC, "Fetching blockfilters from height %d to height %d.\n", start_index->nHeight, block->nHeight);
```
I think this is worth logging in general rather than RPC only, just like `rescanwallet`.

jonatack commented on src/rpc/blockchain.cpp:
```diff
@@ -2284,6 +2284,108 @@ static RPCHelpMan scantxoutset()
+                        {RPCResult::Type::STR_HEX, "", "The blockhash"},
+                    }},
+                RPCExamples{
+                    HelpExampleCli("scanblockfilters", "[\"addr(bcrt1q4u4nsgk6ug0sqz7r3rj9tykjxrsl0yy4d0wwte)\"] 300000") +
```
Looks like the example is missing quotes
```suggestion
                    HelpExampleCli("scanblockfilters", "'[\"addr(bcrt1q4u4nsgk6ug0sqz7r3rj9tykjxrsl0yy4d0wwte)\"]' 300000") +
```
```
$ bitcoin-cli -signet scanblockfilters ["addr(tb1qdydhyacp7tj2asc0u4mhtf6rn65dfjxclrdd7a)"] 30
error: Error parsing JSON: [addr(tb1qdydhyacp7tj2asc0u4mhtf6rn65dfjxclrdd7a)]
```
```
$ bitcoin-cli -signet scanblockfilters '["addr(tb1qdydhyacp7tj2asc0u4mhtf6rn65dfjxclrdd7a)"]' 0
[
  "0000013b318daf6e8d4c531b8293ebf800871f4220fc4b7b25cc6798ec5cb155",
  "000001003f625cbecd53924713929b5a91b12edf4eb844f33781ecb6bb78e69b",
  "000000840904ea8ff963d4f776c0175a622228738cdf9ebd481e3ef483ccdff0"
]

jonasschnelli replied:
Fixed

jonatack commented on src/rpc/blockchain.cpp:
```diff
@@ -2284,6 +2284,108 @@ static RPCHelpMan scantxoutset()
+                            },
+                        },
+                        "[scanobjects,...]"},
+                    {"startheight", RPCArg::Type::NUM, /*default*/ "0", "height to start to filter from"},
```
The actual default start height currently seems to be the tip, not 0.
```cpp
block = ::ChainActive().Tip();
```

jonatack commented on src/rpc/blockchain.cpp:
```diff
@@ -2284,6 +2284,108 @@ static RPCHelpMan scantxoutset()
+{
+    std::string filtertype_name = "basic";
+    if (!request.params[2].isNull()) {
+        filtertype_name = request.params[1].get_str();
```
Looks like this should be `params[2]`
```suggestion
        filtertype_name = request.params[2].get_str();
```

In the functional test, suggest asserting that two calls like this return the same result:
```
bitcoin-cli -signet scanblockfilters \
  '["addr(tb1qdydhyacp7tj2asc0u4mhtf6rn65dfjxclrdd7a)"]' 0 basic
[
  "0000013b318daf6e8d4c531b8293ebf800871f4220fc4b7b25cc6798ec5cb155",
  "000001003f625cbecd53924713929b5a91b12edf4eb844f33781ecb6bb78e69b",
  "000000840904ea8ff963d4f776c0175a622228738cdf9ebd481e3ef483ccdff0"
]

bitcoin-cli -signet scanblockfilters \
  '["addr(tb1qdydhyacp7tj2asc0u4mhtf6rn65dfjxclrdd7a)"]' 0 
[
  "0000013b318daf6e8d4c531b8293ebf800871f4220fc4b7b25cc6798ec5cb155",
  "000001003f625cbecd53924713929b5a91b12edf4eb844f33781ecb6bb78e69b",
  "000000840904ea8ff963d4f776c0175a622228738cdf9ebd481e3ef483ccdff0"
]
```

jonatack replied:
Can be `const`
```diff
- std::string filtertype_name = "basic";
- if (!request.params[2].isNull()) {
-     filtertype_name = request.params[1].get_str();
- }
+ const std::string filtertype_name{request.params[2].isNull() ? "basic" : request.params[2].get_str()};
```

jonasschnelli replied:
Nice. Will adapt this.

jonatack reviewed (commented):
Tested concept ACK

Naming-wise, `scanobjects`, `startheight` and `filtertype` should probably be snakecase.

luke-jr commented:
Seems like block filters are just a potential implementation detail/optimisation here, not part of the fundamental concept...

Maybe rename it to `scanblocks` (and consider supporting a slow scan when the filters are disabled?)

fjahr commented:
Concept ACK

jonasschnelli commented:
> Maybe rename it to scanblocks (and consider supporting a slow scan when the filters are disabled?)

I can do that. But the call doesn't scan blocks (hence your proposed name `scanblocks`). It looks for relevant block**hashes** based on descriptors.
`scanblocks` doesn't sound after what this call is doing.

flack commented:
maybe call it `findblockhashes` then? Or `findblocks` even?

Sjors commented:
I think `scanblocks` is fine. Without filters it "scans blocks" and returns block hashes. With filter it doesn't scan raw block data, but it scans filters block by block.

jonasschnelli commented:
> Without filters it "scans blocks" and returns block hashes. With filter it doesn't scan raw block data, but it scans filters block by block.

Wouldn't it be inconsistent to have `scanblocks` (general) in relation to `rescanblockchain` (wallet)?
But yes... using `scanblock` perhaps is acceptable.

felipsoarez reviewed (commented):
Concept ACK, not tested

jonasschnelli committed:
Add scanblocks RPC call - scan for relevant blocks with descriptors

jonasschnelli committed:
Add scanblock functional test

jonasschnelli commented:
Overhauled this PR. Fixed all the reported issues.

Added the same interface then `scantxoutset` (with action "start", "abort", "status"). Allows to get progress reports via a different RPC call (or abort the current scan). No concurrent scans possible.

Also added `stop_height`.

Some performance reports on an `i7-8700 CPU @ 3.20GHz` with NVme disk:
* Scanning an pkh xpub (`m/0/*` & `m/1/*`, range 0-220 == 440 sPUk) on mainnet from **genesis to tip**: `2m19.176s`
* Same xpub, scan *only* the **last three years**: `1m0.729s`
* Same xpub, "catch up"-scan the **last 5 weeks** (scan the 5040 latest blocks): `0m2.465s`

Single sPUk: 
* Scanning an single address on mainnet from **genesis to tip**: `1m20.258s`
* Same address, scan *only* the **last three years**: `0m41.519s`
* Same address, "catch up"-scan the **last 5 weeks**: `0m1.744s`

Sjors commented:
Thanks for the update! Unfortunately unlike `rescanblockchain` we actually need the result of this call. Just having `status` is not enough to obtain the result if RPC times out. I think we need to store the result somewhere too, perhaps in a way that you can only read once. However if that's too difficult we could instead recommend that the user pays attention to the `start_height` and `stop_height` and calls the RPC in batches for very old wallets.

jonasschnelli commented:
> Thanks for the update! Unfortunately unlike rescanblockchain we actually need the result of this call. Just having status is not enough to obtain the result if RPC times out.

I think we can handle this the same way as `scantxoutset` (both call are in the same group of how much time they consume).
You just need to set the rpctimeout correctly (as one has to do for the three `waitfor*` calls).

I thought about adding dispatched threads. `start` could return a UUID where `status` and `abort` would take this as an argument. But it seems overengineering this.

Let's just keep it identical to `scantxoutset` (they are pretty much identical, one scan the blocks, the other the utxo set).

Sjors commented:
Having long running tasks that you can manage via a UUID does sound useful, but I agree it would be overkill in this PR.

jonasschnelli commented:
Rebased

fjahr reviewed (commented):
Concept ACK

Will do some manual testing soon.

jonasschnelli commented on src/rpc/blockchain.cpp:
```diff
@@ -2290,6 +2290,198 @@ static RPCHelpMan scantxoutset()
+{
+    UniValue ret(UniValue::VOBJ);
+    if (request.params[0].get_str() == "status") {
+        BlockFiltersScanReserver reserver;
```
Could.. yes. But since there is no `else` (unknown action), I tend to keep it at that level.

jonasschnelli commented on src/rpc/blockchain.cpp:
```diff
@@ -2290,6 +2290,198 @@ static RPCHelpMan scantxoutset()
+        const CBlockIndex* stop_block = nullptr;
+        {
+            LOCK(cs_main);
+            block = ::ChainActive().Genesis();
+            stop_block = ::ChainActive().Tip();
```
Yes. That would be possible. But IMO it makes the code hard(er) to read and I guess performance wise is it almost no impact.

jonasschnelli commented on src/rpc/blockchain.cpp:
```diff
@@ -2290,6 +2290,198 @@ static RPCHelpMan scantxoutset()
+        }
+        CHECK_NONFATAL(block);
+
+        // loop through the scan objects, add scripts to the element_set
```
Good catch. Fixed.

jonasschnelli commented on src/rpc/blockchain.cpp:
```diff
@@ -2290,6 +2290,198 @@ static RPCHelpMan scantxoutset()
+            block = next;
+        }
+        ret.pushKV("from_height", start_block->nHeight);
+        ret.pushKV("to_height", last_scanned_block->nHeight);
```
Not sure about this. The interruption point is at the beginning of the while loop. Processing blocks (scanning filters) can't be done without setting `last_scanned_block = block;`. Shouldn't therefore `last_scanned_block` always point to `stopindex` of `LookupFilterRange()` (a.k.a. the last scanned block)? Or do I miss something?

fjahr replied:
Sorry, that I didn't get around to answer earlier. I am trying to simplify so please ignore if I get some minor details wrong: Let's say we are scanning blocks 10.000 - 30.000. The while loop runs through every block and with each block `last_scanned_block` is updated. But no scanning is actually happening. At 20.000 it does the actual scanning for the first time, blocks 10k-20k. So at height 20k `last_scanned_block` is actually in sync with that, up to 20k was scanned and the `last_scanned_block` is 20k. From the next block on `last_scanned_block` increments with every loop again and is out of sync with what was actually synced. So if the interruption point is hit at, for example, 25k, then `to_height` would say 25k but actually, the scan was only done up to 20k. One very simple fix would be to move `last_scanned_block = block;
` just 2 lines higher so it is only updated after an actual scan. But then it is just reports the same as `g_scanfilter_progress_height` so I think it can be dropped altogether. I don't see `last_scanned_block` being used anywhere aside for the `to_height` return value usage in this line here, so I am not sure what you mean by "it can't be done without it". And just fundamentally I don't see why this would not be the same: The progress height and the last scanned block are the same value to me.

I just realized I am writing a whole paragraph for a 3 line change, code is probably much easier to reason about :) https://github.com/fjahr/bitcoin/commit/16c7cb5b67e95ba1c8050eacfd4099fad108acf3 Tests pass.

achow101 replied:
In 6a69dd267e4b96e6657de7b77521311994b2a902 "Add scanblocks RPC call - scan for relevant blocks with descriptors"

I agree with @fjahr here. `last_scanned_block` isn't actually the last block we have scanned, and if we abort, then `to_height` does not accurately reflect where we have actually scanned up to. It would be better to move it up into the `if` the guards the actual scanning. But in that case, it's the same as `g_scanfilter_progress_height`, so it can just be removed.

jamesob replied:
Fixed in #23549

jonasschnelli commented on src/rpc/blockchain.cpp:
```diff
@@ -2290,6 +2290,198 @@ static RPCHelpMan scantxoutset()
+                next = ChainActive().Next(block);
+                if (block == stop_block) next = nullptr;
+            }
+            if (start_index->nHeight+amount_per_chunk == block->nHeight || next == nullptr) {
```
Fixed.

jonasschnelli commented on src/rpc/blockchain.cpp:
```diff
@@ -2290,6 +2290,198 @@ static RPCHelpMan scantxoutset()
+                    {"action", RPCArg::Type::STR, RPCArg::Optional::NO, "The action to execute\n"
+            "                                      \"start\" for starting a scan\n"
+            "                                      \"abort\" for aborting the current scan (returns true when abort was successful)\n"
+            "                                      \"status\" for progress report (in %) of the current scan"},
```
Good point. Added.

jonasschnelli commented:
Thanks @fjahr for the review. Fixed the reported points (see also answer https://github.com/bitcoin/bitcoin/pull/20664#discussion_r578245439).

Sjors commented:
CI is unhappy

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -2288,6 +2288,198 @@ static RPCHelpMan scantxoutset()
+{
+    return RPCHelpMan{"scanblocks",
+                "\nReturn relevant blockhashes for given descriptors.\n"
+                "This call may take serval minutes. Make sure to use no RPC timeout (bitcoin-cli -rpcclienttimeout=0)",
```
Maybe add a hint here that `-debug=rpc` will show progress in the log (this behavior is different from `rescanblockchain` which always prints to the log)

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -2288,6 +2288,198 @@ static RPCHelpMan scantxoutset()
+        ret.pushKV("relevant_blocks", blocks);
+    }
+    else {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid command");
```
nit: "Invalid action argument"

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -2288,6 +2288,198 @@ static RPCHelpMan scantxoutset()
+                        // compare the elements-set with each filter
+                        if (filter.GetFilter().MatchAny(needle_set)) {
+                            blocks.push_back(filter.GetBlockHash().GetHex());
+                            LogPrint(BCLog::RPC, "scanblocks: found match in %s\n", filter.GetBlockHash().GetHex());
```
Maybe add ` (height)`

jonasschnelli replied:
Would require to fetch the height from the hash (just for the sake of logging).

Sjors reviewed (commented):
73077f1b795d1b3325896b01018de63fcf700ddd looks mostly good to me. I was able to test as well with some descriptors.

jonasschnelli committed:
Unify "invalid action argument" error (fix scantxoutset)

jonasschnelli commented:
Fixed the RPC test failure and implemented some of @Sjors suggestions.

chrisguida commented:
ACK 71b7cdb460e2d0179aa87bdd0d30a82a821f6d05

I reviewed the code, ran the functional tests, and did some manual testing. This rpc call works great; I was able to scan 1000 addresses from genesis to tip on an M1 Macbook Air in 3 min 27 sec.

The only nit I have is that the `bitcoin-cli help scanblocks` examples don't work, but the functions that create the examples (`HelpExampleCli` and `HelpExampleRpc`) don't seem to accept an `action` argument, and I'm not sure what adding one would break. The other problem is that the examples use double quotes around the json list of scan objects, when the rpc api is expecting single quotes, as in the example in the comment at the top of this PR (that one works!).

I'm super excited for this feature to get merged, as it will enable wallet trackers such as EPS and BWT to be practical with pruned nodes. Is it possible for this to be released in 0.22?

Thanks @jonasschnelli for making this happen!

Sjors commented on test/functional/rpc_scanblockfilters.py:
```diff
@@ -0,0 +1,73 @@
+        # mine a block and assure that the mined blockhash is in the filterresult
+        blockhash = self.nodes[0].generate(1)[0]
+        out = self.nodes[0].scanblocks("start", ["addr("+addr_1+")"])
+        assert(blockhash in out['relevant_blocks'])
```
This assertion fails frequently on my machine.

Sjors commented on test/functional/rpc_scanblockfilters.py:
```diff
@@ -0,0 +1,73 @@
+        # make sure the blockhash is not in the filter result if we set the start_height to the just mined block (unlikely to hit a false positive)
+        assert(blockhash not in self.nodes[0].scanblocks("start", ["addr("+addr_1+")"], self.nodes[0].getblockheader(blockhash_new)['height'])['relevant_blocks'])
+
+        # make sure the blockhash is present when using the first mined block as start_height
```
When the above assertion doesn't fail, then this one tends to fail.

Sjors reviewed (commented):
71b7cdb460e2d0179aa87bdd0d30a82a821f6d05 still looks good, but I'm getting frequent test failures on my machine. Frequent, but not always, sometimes the test does pass.

First I thought it was due to me using descriptor wallets (`./configure --without-bdb` does that). But I added a sanity check for that: https://github.com/Sjors/bitcoin/commit/44c6b38b2878c1c559dd9b1128b978d9e66afe7e

Also, fjahr@16c7cb5's suggestion makes sense to me.

chrisguida commented on src/rpc/blockchain.cpp:
```diff
@@ -2288,6 +2288,198 @@ static RPCHelpMan scantxoutset()
+        }
+        CHECK_NONFATAL(block);
+
+        // loop through the scan objects, add scripts to the needle_set
```
Was thinking this, I'm guessing we'll want it abstracted out if we want to be able to perform internal rescans, if, for example, the user decides to add an old wallet.

Sjors replied:
It would also help with GUI support. I image soon (tm) to have a wizard where you can import a BIP39 seed and it recovers coins.

jamesob replied:
I'd like to handle moving this code out of RPC land in a future change. It isn't a trivial move (given the global progress state), people are probably going to want unittests if we do the move, and IMO the important short-term objective here is getting fast rescans working ASAP. So if it's okay with you guys I'd like to save this move for a later PR once.

prayank23 commented:
Concept ACK. Will test today.

Two questions:

1. Can we include -rpctimeout=0 by default if not mentioned by the user because the call can take time?
2. Do we have any RPC which scans mempool instead of blocks and return the transactions associated with it?

prayank23 commented:
Compiled successfully on Ubuntu. Two functional tests failed:

```
feature_config_args.py                             | ✖ Failed  | 34 s
feature_pruning.py                                 | ✖ Failed  | 480 s
```

MarcoFalke commented:
What is the combined log of the test failure? `./test/functional/combine_logs.py`

kiminuo commented on src/rpc/blockchain.cpp:
```diff
@@ -2281,13 +2281,205 @@ static RPCHelpMan scantxoutset()
+
+                // update progress
+                int blocks_processed = block->nHeight - start_block->nHeight;
+                int total_blocks_to_process = stop_block->nHeight - start_block->nHeight;
```
nit: It seems to me that this line can be put out of `while` cycle.

jamesob replied:
Fixed in #23549

kiminuo commented on src/rpc/blockchain.cpp:
```diff
@@ -2281,13 +2281,205 @@ static RPCHelpMan scantxoutset()
+        while (block) {
+            node.rpc_interruption_point(); // allow a clean shutdown
+            if (g_scanfilter_should_abort_scan) {
+                break;
```
Would it be helpful to log that the scanning was aborted?

jamesob replied:
Fixed in #23549

kiminuo commented on src/rpc/blockchain.cpp:
```diff
@@ -2281,13 +2281,205 @@ static RPCHelpMan scantxoutset()
+                            },
+                        },
+                        "[scanobjects,...]"},
+                    {"start_height", RPCArg::Type::NUM, /*default*/ "0", "height to start to filter from"},
```
```suggestion
                    {"start_height", RPCArg::Type::NUM, /*default*/ "0", "Height to start to filter from"},
```
?

Is `filter` correct verb here? `scan`?

jamesob replied:
Fixed in #23549

kiminuo commented on src/rpc/blockchain.cpp:
```diff
@@ -2281,13 +2281,205 @@ static RPCHelpMan scantxoutset()
+                        },
+                        "[scanobjects,...]"},
+                    {"start_height", RPCArg::Type::NUM, /*default*/ "0", "height to start to filter from"},
+                    {"stop_height", RPCArg::Type::NUM, /*default*/ "<tip>", "height to stop to scan"},
```
Similar to above.

kiminuo commented on src/rpc/blockchain.cpp:
```diff
@@ -2281,13 +2281,205 @@ static RPCHelpMan scantxoutset()
+{
+    return RPCHelpMan{"scanblocks",
+                "\nReturn relevant blockhashes for given descriptors.\n"
+                "This call may take serval minutes. Make sure to use no RPC timeout (bitcoin-cli -rpcclienttimeout=0)",
```
```suggestion
                "This call may take several minutes. Make sure to use no RPC timeout (bitcoin-cli -rpcclienttimeout=0)",
```

jamesob replied:
Fixed in https://github.com/bitcoin/bitcoin/pull/23549

kiminuo commented on src/rpc/blockchain.cpp:
```diff
@@ -2281,13 +2281,205 @@ static RPCHelpMan scantxoutset()
+
+        // loop through the scan objects, add scripts to the needle_set
+        GCSFilter::ElementSet needle_set;
+        for (const UniValue& scanobject : request.params[1].get_array().getValues()) {
```
For my information: This array seems to unbounded. The bound is probably some maximum size of RPC requests in general. Anyway, I wonder whether it is an issue or not. I guess not because nobody mentioned it here. Could anybody elaborate on this a bit, please?

jamesob replied:
This is only subject to authenticated RPC user input and isn't handled in previous RPC calls (e.g. `scantxoutset`) so if someone wants to DoS their own node with a massive input array that seems okay to me.

kiminuo commented on src/rpc/blockchain.cpp:
```diff
@@ -2281,13 +2281,205 @@ static RPCHelpMan scantxoutset()
+        std::vector<BlockFilter> filters;
+        const CBlockIndex* start_block = block; // for progress reporting
+        const CBlockIndex* last_scanned_block = block;
+        g_scanfilter_should_abort_scan = false;
```
For my information: Would it make sense to move this line to L2414 and to make the for-loop on L2417 abortable too? Or is it an overkill? I'm not sure how long the for loop can take for "many scanobjects".

jamesob replied:
This is overkill IMO; even for a long list, `EvalDescriptorStringOrObject` should be quite fast since it doesn't rely on an disk IO, unlike the calls below.

kiminuo commented on test/functional/rpc_scanblockfilters.py:
```diff
@@ -0,0 +1,73 @@
+# Copyright (c) 2021 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test the scanblocks rpc call."""
```
nit: What about:

```suggestion
"""Test the scanblocks RPC call."""
```
?

It looks like it is mostly written in this way in tests.

jamesob replied:
Fixed in #23549

kiminuo commented on test/functional/rpc_scanblockfilters.py:
```diff
@@ -0,0 +1,73 @@
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal, assert_raises_rpc_error
+
+class scanblocksTest(BitcoinTestFramework):
```
Nit:

```suggestion
class ScanblocksTest(BitcoinTestFramework):
```

jamesob replied:
Fixed in #23549

achow101 commented on src/rpc/blockchain.cpp:
```diff
@@ -2288,6 +2288,198 @@ static RPCHelpMan scantxoutset()
+                    {"stop_height", RPCArg::Type::NUM, /*default*/ "<tip>", "height to stop to scan"},
+                    {"filtertype", RPCArg::Type::STR, /*default*/ "basic", "The type name of the filter"}
+                },
+                RPCResult{
```
In 6a69dd267e4b96e6657de7b77521311994b2a902 "Add scanblocks RPC call - scan for relevant blocks with descriptors"

This `RPCResult` does not reflect the actual results.

jamesob replied:
Fixed in #23549

achow101 commented on test/functional/rpc_scanblockfilters.py:
```diff
@@ -0,0 +1,73 @@
+        self.nodes[0].sendtoaddress("mkS4HXoTYWRTescLGaUTGbtTTYX5EjJyEE", 1.0) # childkey 5 of tpubD6NzVbkrYhZ4WaWSyoBvQwbpLkojyoTZPRsgXELWz3Popb3qkjcJyJUGLnL4qHHoQvao8ESaAstxYSnhyswJ76uZPStJRJCTKvosUCJZL5B
+
+        # mine a block and assure that the mined blockhash is in the filterresult
+        blockhash = self.nodes[0].generate(1)[0]
```
In 71b7cdb460e2d0179aa87bdd0d30a82a821f6d05 "Add scanblock functional test"

Since the indexes are built asynchronously, it is possible that after generating the block, the block filter index has not updated to include the new block, so scanblocks might not find anything. Instead, we need to wait for it to become synced:

```suggestion
        blockhash = self.nodes[0].generate(1)[0]
        self.wait_until(lambda: all(i["synced"] for i in self.nodes[0].getindexinfo().values()))
```

jamesob replied:
Fixed in #23549

theStack commented:
Concept ACK

There have been various code reviews but no changes since months. @jonasschnelli: are you still working on this PR? (Just checking out before I start to code-review).

MarcoFalke commented:
Removed from high-prio for now, but can be added back any time.

Kirandevraj commented:
Compiled and tested Successfully 6a69dd267e4b96e6657de7b77521311994b2a902 at Ubuntu 18.04. 
Started Bitcoin Core in regtest with keypool=5. Created and mined blocks to addresses in descriptor wallet after creating addresses more than the keypool size using getnewaddress. Scanned blocks for all the addresses using `scanblocks` RPC and were able to get the right blocks that contained the transactions in it for all the addresses.

darosior reviewed (commented):
Concept ACK

jamesob commented:
Concept ACK. Would love to see this, or any other mechanism that allows performing wallet rescans in ~2min on good hardware, make it in.

luke-jr commented:
Rebased here: https://github.com/bitcoin/bitcoin/compare/master...luke-jr:rpc_filterblocks

MarcoFalke commented:
Should this be marked "up for grabs" @jonasschnelli ?

jamesob commented:
I haven't started review yet, but given @MarcoFalke has done some previous work on this, we should ensure that this change isn't subject to the bugs he points out here: https://github.com/bitcoin/bitcoin/pull/15845#issuecomment-625768563

jamesob commented:
For what it's worth, I'm working on resurrecting this PR. Hoping to have a branch pushed by end of day.

jamesob commented on test/functional/rpc_scanblockfilters.py:
```diff
@@ -0,0 +1,73 @@
+        # also test the stop height
+        assert(blockhash in self.nodes[0].scanblocks("start", ["addr("+addr_1+")"], self.nodes[0].getblockheader(blockhash)['height'], self.nodes[0].getblockheader(blockhash)['height'])['relevant_blocks'])
+
+        # use the stop_height to exclude the relevent block
```
Fixed in #23549

meshcollider commented:
Please see #23549
